src/middlewared/middlewared/plugins/jbof/crud.py

```diff
@@ -0,0 +1,546 @@
+        previously determined subnets.
+
+        Then attempt to connect using all the available RDMA capable
+        interfaces.
+        """
+        # shelf_interfaces = await self.middleware.call('jbof.fabric_interface_choices', mgmt_ip)
+        await self.middleware.call('jbof.hardwire_shelf', mgmt_ip, shelf_index)
+        await self.middleware.call('jbof.hardwire_host', mgmt_ip, shelf_index, schema, verrors)
+        if not verrors:
+            await self.middleware.call('jbof.attach_drives', mgmt_ip, shelf_index, schema, verrors)
+
+    @private
+    def fabric_interface_choices(self, mgmt_ip):
+        redfish = RedfishClient.cache_get(self.middleware, mgmt_ip)
+        return redfish.fabric_ethernet_interfaces()
+
+    @private
+    def fabric_interface_macs(self, mgmt_ip):
+        """Return a dict keyed by IP address where the value is the corresponding MAC address."""
+        redfish = RedfishClient.cache_get(self.middleware, mgmt_ip)
+        macs = {}
+        for uri in self.fabric_interface_choices(mgmt_ip):
+            netdata = redfish.get_uri(uri)
+            for address in netdata['IPv4Addresses']:
+                macs[address['Address']] = netdata['MACAddress']
+        return macs
+
+    @private
+    def hardwire_shelf(self, mgmt_ip, shelf_index):
+        redfish = RedfishClient.cache_get(self.middleware, mgmt_ip)
+        shelf_interfaces = redfish.fabric_ethernet_interfaces()
+        for (eth_index, uri) in enumerate(shelf_interfaces):
+            address = jbof_static_ip(shelf_index, eth_index)
+            redfish.configure_fabric_interface(uri, address, static_ip_netmask_str(address), mtusize=static_mtu())
+        time.sleep(20)
+
+    @private
+    def unwire_shelf(self, mgmt_ip):
+        redfish = RedfishClient.cache_get(self.middleware, mgmt_ip)
+        for uri in redfish.fabric_ethernet_interfaces():
+            redfish.configure_fabric_interface(uri, '0.0.0.0', '255.255.255.0', True, mtusize=1500)
+
+    @private
+    async def hardwire_host(self, mgmt_ip, shelf_index, schema, verrors):
+        """Discover which direct links exist to the specified expansion shelf."""
+        # See how many interfaces are available on the expansion shelf
+        shelf_interfaces = await self.middleware.call('jbof.fabric_interface_choices', mgmt_ip)
+        shelf_ip_to_mac = await self.middleware.call('jbof.fabric_interface_macs', mgmt_ip)
+        shelf_ips = list(shelf_ip_to_mac.keys())
+
+        # Setup a dict with the expected IP pairs
+        shelf_ip_to_host_ip = {}
+        for eth_index in range(0, len(shelf_interfaces)):
+            shelf_ip_to_host_ip[jbof_static_ip(shelf_index, eth_index)] = initiator_static_ip(shelf_index, eth_index)
+
+        # Let's check that we have the expected hardwired IPs on the shelf
+        if set(shelf_ips) != set(list(shelf_ip_to_host_ip.keys())):
+            # This should not happen
+            verrors.add(schema, 'JBOF does not have expected IPs.')
```

Comment:
Would it be helpful to admin to include what IPs were expected and which ones were actually set?